### PR TITLE
Fix completion of paths

### PR DESCRIPTION
**Pull request by matklad**
r? @flodiebold

### CODE REVIEW

File: crates/ra_hir/src/resolve.rs [hunk 1]
```diff
@@ -56,7 +56,7 @@ pub enum Resolution {
 }
 
 impl Resolver {
-    pub fn resolve_name(&self, name: &Name) -> PerNs<Resolution> {
+    fn resolve_name(&self, name: &Name) -> PerNs<Resolution> {
```
Comment (matklad) [inline crates/ra_hir/src/resolve.rs]: Given that it's unclear if this looks only at the scopes or in the item_map as well, it seems to be a good idea to make this private.

File: crates/ra_ide_api/src/completion/complete_path.rs [hunk 2]
```diff
@@ -73,7 +73,7 @@ mod tests {
     }
 
     #[test]
-    fn completes_mod_with_docs() {
+    fn complets_mod_with_docs() {
```
Comment (flodiebold) [inline crates/ra_ide_api/src/completion/complete_path.rs]: wasn't `completes` the correct spelling?

File: crates/ra_hir/src/resolve.rs (hunk 1 above)
Comment (flodiebold) [reply crates/ra_hir/src/resolve.rs]: it should be looking at the item map as well (if there's a `ModuleScope` on the stack), I think...

File: crates/ra_ide_api/src/completion/complete_path.rs (hunk 2 above)
Comment (matklad) [reply crates/ra_ide_api/src/completion/complete_path.rs]: :sweat_smile:

File: crates/ra_hir/src/resolve.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_hir/src/resolve.rs]: Hm, so the bug is that we don't have a module scope active? Will look into this, thanks!

**Comment by matklad** [comment]
@flodiebold how this looks now?

File: crates/ra_hir/src/nameres.rs [hunk 3]
```diff
@@ -434,6 +434,14 @@ impl ItemMap {
         self.resolve_path_fp(db, original_module, path).0
     }
 
+    pub(crate) fn resolve_name_in_module(&self, module: Module, name: &Name) -> PerNs<ModuleDef> {
+        let from_scope = self[module.module_id].items.get(name).map_or(PerNs::none(), |it| it.def);
+        let from_extern_prelude =
+            self.extern_prelude.get(name).map_or(PerNs::none(), |&it| PerNs::types(it));
+
+        from_scope.combine(from_extern_prelude)
```
Comment (flodiebold) [inline crates/ra_hir/src/nameres.rs]: maybe there should be a better name than `combine` that makes clear what shadows what :smile:

File: crates/ra_hir/src/nameres.rs [hunk 4]
```diff
@@ -434,6 +434,14 @@ impl ItemMap {
         self.resolve_path_fp(db, original_module, path).0
     }
 
+    pub(crate) fn resolve_name_in_module(&self, module: Module, name: &Name) -> PerNs<ModuleDef> {
+        let from_scope = self[module.module_id].items.get(name).map_or(PerNs::none(), |it| it.def);
+        let from_extern_prelude =
+            self.extern_prelude.get(name).map_or(PerNs::none(), |&it| PerNs::types(it));
```
Comment (flodiebold) [inline crates/ra_hir/src/nameres.rs]: I always forget that `map_or` exists

**Comment by flodiebold** [comment]
Looks good to me!

bors r+

File: crates/ra_hir/src/nameres.rs (hunk 4 above)
Comment (matklad) [reply crates/ra_hir/src/nameres.rs]: Well, one day we'll implement a "weak_warning" for this... :-)

File: crates/ra_hir/src/nameres.rs (hunk 3 above)
Comment (matklad) [reply crates/ra_hir/src/nameres.rs]: `.or`?

**Pull request merged**